src/sentry/processing/backpressure/monitor.py

```diff
@@ -0,0 +1,107 @@
+import time
+from dataclasses import dataclass
+from typing import Dict, Generator, List, Mapping, Union
+
+from django.conf import settings
+
+from sentry import options
+from sentry.processing.backpressure.health import record_consumer_heath
+from sentry.processing.backpressure.rabbitmq import (
+    RabbitMqHost,
+    parse_rabbitmq,
+    query_rabbitmq_memory_usage,
+)
+
+# from sentry import options
+from sentry.processing.backpressure.redis import Cluster, iter_cluster_memory_usage
+from sentry.utils import redis
+
+
+@dataclass
+class Redis:
+    cluster: Cluster
+
+
+@dataclass
+class RabbitMq:
+    servers: List[RabbitMqHost]
+
+
+Service = Union[Redis, RabbitMq]
+
+
+@dataclass
+class ServiceMemory:
+    used: int
+    available: int
+    percentage: float
+
+    def __init__(self, used: int, available: int):
+        self.used = used
+        self.available = max(available, 1)
```

Comment:
The contract is not clear if `available` is passed as `0`. You will end up with a percentage of `1.0` (`min(large/1, 1) = 1`), which will further down mark the service as unhealthy (1.0 >= some_watermark).

Your implementation in Redis currently falls back from `maxmemory` to `total_system_memory`, which is a brittle assumption to make here, especially since that code also falls back to `0` again. Suggestions:
1. Set `percentage` to `0` if there's no available memory
2. Set `percentage` to `None` and handle that case in the caller. That's slightly more semantic, since we actually don't know anything about saturation.
3. Deny `0` and raise an error here; ensure outside that this is never constructed.

Either way, to be defensive and place intent, I'd suggest to:
 - Describe semantics in a doc comment on the class or initializer
 - Consider adding an explicit check (`if available == 0`) so that this is baked into the implementation.